Revscriptsys new design

EvilHero90 opened the pull request:
<!-- Note: Lines with this <!-- syntax are comments and will not be visible in
     your pull request. You can safely ignore or remove them. -->

### Pull Request Prelude

<!-- Thank you for working on improving The Forgotten Server! -->
<!-- Please complete these steps and check the following boxes by putting an `x`
     inside the [brackets] before filing your Pull Request. -->

- [x] I have followed [proper The Forgotten Server code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR may be closed if the above-mentioned criteria are not fulfilled.

### Changes Proposed

- We are now able to register an Event object directly without the need to call `:register`
- Another thing which is changed is that you can name the hook function individual now ex: `onUseShovel` incase the script errors it shows this exact event name.
This is how it works now
```lua
local test = Action({id={100,2000,300}, allowFarUse=true})
function test.nameThisWhateverYouWant(...)
    return true
end
```
I'll obviously re write all of the other interfaces aswell.
The aim is to be able to hook the actions back and be able to overwrite them (makes /reload redundant)

<!-- Describe the changes that this pull request makes. -->


<!-- You can safely ignore the links below:  -->

[cont]: https://github.com/otland/forgottenserver/wiki/Contributing
[code]: https://github.com/otland/forgottenserver/wiki/TFS-Coding-Style-Guide

EvilHero90 committed:
Action interface rework

We are now able to register an Action object directly without the need to call `:register`
Another thing which is changed is that you can name the hook function individual now ex: `onUseShovel` incase the script errors it shows this exact event name.

EvilHero90 committed:
compiler fix

EvilHero90 committed:
small fix

forgot to add a default return value of a `nullptr` incase it doesn't find the action we search for

EvilHero90 committed:
same mistake

wrong reference

EvilHero90 committed:
hook for event

We can now hook back into the event and overwrite the lua function, we are removing the old one out of the cacheFiles to avoid memory leaks and then set scriptId to 0 to let it asign a new scriptId

EvilHero90 committed:
clang...

EvilHero90 committed:
Merge branch 'master' into revscriptsys-new

EvilHero90 committed:
working Actions now

this version is now smoothly working for Actions

EvilHero90 committed:
TalkAction

works as a charm now aswell

EvilHero90 committed:
Action* to Action_shared_ptr

- getting rid of raw pointing, now it wont crash if the setup isn't done right
- the shared_ptr actually never gets null as the raw* is saved in the map, so we are not leaking memory as it was expected before, we are only freeing at /reload this also makes sure there are no crashes at reload, also introducing the __gc function.
- we are now able to make Action as a global variable and access it from everywhere else later if needed (I would still keep it local if data is static)

EvilHero90 committed:
TalkAction* to TalkAction_shared_ptr

same as Action

EvilHero90 committed:
clang

EvilHero90 committed:
string to const string&

EvilHero90 committed:
optimizations

- re work on the talkaction register function, we are now displaying if we accidently used a command word twice

EvilHero90 committed:
MoveEvent* into shared pointer

same as action and talkaction.

EvilHero90 committed:
clang...

EvilHero90 committed:
added Game.getMoveEvent & fixes

- added Game.getMoveEvent
- made a few fixes with references
- exposed MOVE_EVENT enums to lua

EvilHero90 committed:
Update actions.cpp

EvilHero90 committed:
Update movement.cpp

EvilHero90 committed:
CreatureEvent* to shared_pointer

same as the other ones

EvilHero90 committed:
GlobalEvent* to shared_ptr

- this also includes a fix in creatureevent.lua a variable name was incorrect
- fixes also the onTime event, this was somehow not correctly set and no one ever noticed that
- merged all the different globalevent maps together, we name them unique anyway, so it makes no sense for different maps, this brings in a bit more simplicity

EvilHero90 committed:
clang... as usual...

EvilHero90 committed:
memory leak fix

I've accidently created a memory leak which didn't clear the cacheFiles correctly, now instead of clearing the cacheFiles we just re use them on reload

EvilHero90 committed:
fixed params for the lua files

EvilHero90 committed:
Weapon* to shared_ptr

same as the others
providing an example case with burst arrow

EvilHero90 committed:
MoveEvent and Weapon additions for registering

- I've enabled that MoveEvents or Weapons can no be registered without function body, for default core functionality
- added snakebite_rod.lua as an example

EvilHero90 committed:
added some checks to not double register

EvilHero90 committed:
fix for overwrite of same function names

- if we are using default function names like `onSay` we where overwriting eachother as this is our indicator for re using the scriptId, we are saving default function names + their loading file string, that way we don't overwrite them
- we are still not able to place 2 different `onUse` events in the same file, unless we name the function different for each, or we are going to overwrite it again.

EvilHero90 committed:
information of outdated revscriptsys version

I've shortened down the console prints to a minimum, just making aware that there are still scripts in the old format

EvilHero90 committed:
Spell* to shared_ptr

same as the others

EvilHero90 committed:
fix to the `:vocation` function

Now the vocations are properly shown

EvilHero90 committed:
debug information

now shows the file containing an old style revscriptsys setup

EvilHero90 committed:
optimizations

-decreasing strings and merging them already
- normal vectors are now const static
- re named some error informations

EvilHero90 committed:
clang

EvilHero90 committed:
Game.getWeapon(id) & Game.getGlobalEvent(name)

ranisalt commented on src/actions.cpp:
```diff
@@ -301,6 +308,27 @@ Action* Actions::getAction(const Item* item)
+	if (type == "id") {
+		auto it = useItemMap.find(id);
+		if (it != useItemMap.end()) {
+			return Action_shared_ptr(&it->second);
```
Here you should just `return it->second` instead. Right now, `useItemMap` is cleared, then all the shared pointers will leak and point to invalid memory.

You should use [`enable_shared_from_this`](https://en.cppreference.com/w/cpp/memory/enable_shared_from_this) here, which provides the functionality.

EvilHero90 replied:
I guess this is resolved as I converted everything to `shared_ptr`

ranisalt commented on src/actions.cpp:
```diff
@@ -301,6 +308,27 @@ Action* Actions::getAction(const Item* item)
 
+Action_shared_ptr Actions::getActionEvent(const std::string& type, uint16_t id)
+{
+	if (type == "id") {
```
Type is finite, so use an enum with a jump table (switch case)

EvilHero90 committed:
modernizing code

- just adapting to new standards

EvilHero90 committed:
fix compiler error

EvilHero90 committed:
deprecation of the old xml system for Events (Action/TalkAction/etc.)

- finaly got rid of all the old xml dependencies when it comes to Events
- re worked everything to fully work with shared_ptr, we aren't operating with raw pointers for Events now anymore.
- we are moving all the scripts in a separate PR into revscriptsys

EvilHero90 committed:
leftover removal
